### PR TITLE
06765 add duplicate check in handle workflow

### PR DESCRIPTION
**Pull request by vtronkov**
**Description**:
Adding check for duplicated transactions in HandleWorkflow

**Node:** It should be merged after https://github.com/hashgraph/hedera-services/pull/7556 - this PR extends RecordListBuilder so that the right TransactionRecord can be pulled

**Related issue(s)**:

Fixes https://github.com/hashgraph/hedera-services/issues/6765

**Notes for reviewer**:

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/ingest/IngestChecker.java [hunk 1]
```diff
@@ -142,6 +148,20 @@ public TransactionInfo runAllChecks(@NonNull final HederaState state, @NonNull f
...
+        final var foundTransactionRecord = hederaRecordCache.getRecord(txBody.transactionIDOrThrow());
+        if (foundTransactionRecord != null) {
+            if (txBody.nodeAccountID() != null
+                    && txBody.nodeAccountID()
+                            .equals(txBody.transactionIDOrThrow().accountID())) {
+                throw new PreCheckException(DUPLICATE_TRANSACTION);
+            }
+        }
+    }
+
```
Comment (rbair23) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/ingest/IngestChecker.java:164]: I think actually in ingest checker we should just remove the "step 2" TODO. We do deduplication check in SubmissionManager and don't need to do a check here.

File: hedera-node/hedera-app/src/test/java/com/hedera/node/app/workflows/handle/HandleWorkflowTest.java [hunk 2]
```diff
@@ -495,6 +529,30 @@ void testPreHandleWithDueDiligenceFailure() {
...
+        verify(recordManager, times(1)).endUserTransaction(any());
+    }
+
+    private TransactionRecord generateDummyTransactionRecord() {
+        final var body = new OneOf<>(BodyOneOfType.UNSET, 1);
+        final var entropy = new OneOf<>(EntropyOneOfType.PRNG_BYTES, 1);
+        return new TransactionRecord(
+                null, null, null, null, null, 1L, body, null, null, null, null, null, null, null, null, null, entropy,
+                null);
+    }
```
Comment (rbair23) [inline hedera-node/hedera-app/src/test/java/com/hedera/node/app/workflows/handle/HandleWorkflowTest.java:554]: It would be good to add a test that doesn't use a mocked HederaRecordCache (or at least, one that remembers values it was previously called with), and actually run two duplicate transactions and make sure the first one goes through fine, and the second one throws a duplicate exception. This particular test, for example, didn't catch the fact that the code was missing the `add` call, whereas the test I'm suggesting would have caught it.

**Review by rbair23 — CHANGES_REQUESTED** [review]

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java [hunk 3]
```diff
@@ -211,6 +219,13 @@ private void handlePlatformTransaction(
         recordManager.endUserTransaction(recordListBuilder.build());
     }
 
+    private void checkDuplicates(final TransactionBody txBody) throws PreCheckException {
```
[the remark this answers is not in the document]
Comment (vtronkov) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java]: Yes, I think this is a good practice. I was passing the body so I can check if the body is null and then proceed with the validation. Then when I was debugging I saw that the code won't reach my code if the body is null anyway and I removed that checked but forgot to refactor the method. Fixed 👍

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/WorkflowsInjectionModule.java [hunk 4]
```diff
@@ -32,4 +36,9 @@
             PreHandleWorkflowInjectionModule.class,
             QueryWorkflowInjectionModule.class
         })
-public interface WorkflowsInjectionModule {}
+public interface WorkflowsInjectionModule {
+
+    @Binds
+    @Singleton
+    HederaRecordCache bindHederaRecordCache(RecordCacheImpl recordCache);
+}
```
[the remark this answers is not in the document]
Comment (vtronkov) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/WorkflowsInjectionModule.java]: Makse sense!
I added the `HederaRecordCache` mapping together with the `RecordCache` because otherwise the classes that require `RecrodCache` are complaining.

An interesting thing is that under the hood Dagger uses the same recordCacheImplProvider for both methods but if I remove the HederaRecordCache binding it complains. I guess this is how Dagger works 🤷 
```
@Override
public RecordCache recordCache() {
  return recordCacheImplProvider.get();
}

@Override
public HederaRecordCache hederaRecordCache() {
  return recordCacheImplProvider.get();
}
```

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java [hunk 5]
```diff
@@ -211,6 +219,13 @@ private void handlePlatformTransaction(
...
+        final var foundTransactionRecord = hederaRecordCache.getRecord(txBody.transactionIDOrThrow());
+        if (foundTransactionRecord != null) {
+            throw new PreCheckException(DUPLICATE_TRANSACTION);
+        }
```
[the remark this answers is not in the document]
Comment (vtronkov) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java]: I presumed that this was happening somewhere else(I should've checked!).
I added a method adding the transaction to the cache. I'm not 100% sure about that tho:
`final var nodeId = preHandleResult.txInfo().txBody().nodeAccountID().accountNum();`
Not sure if `accountNum` is the same as `nodeId`. Let me know if I need to change it.

[the remark this answers is not in the document]
Comment (vtronkov) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java]: I am also not sure about the way I get the TransactionRecord

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/HederaInjectionComponent.java [hunk 6]
```diff
@@ -92,6 +93,8 @@ public interface HederaInjectionComponent {
 
     RecordCache recordCache();
 
+    HederaRecordCache hederaRecordCache();
```
Comment (rbair23) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/HederaInjectionComponent.java:96]: Do we need these two lines? I think what we have in `HederaStateInjectionModule` should be sufficient?

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java [hunk 7]
```diff
@@ -134,14 +141,20 @@ private void handlePlatformTransaction(
...
             final var preHandleResult = getCurrentPreHandleResult(state, platformEvent, platformTxn, configuration);
             recordBuilder.transaction(
                     preHandleResult.txInfo().transaction(),
                     preHandleResult.txInfo().signedBytes());
 
+            // Check transaction duplication
+            checkDuplicates(preHandleResult.txInfo().txBody().transactionIDOrThrow());
+
+            // If the transaction is not duplicated we want to add it to the cache
+            addTransactionToCache(preHandleResult, recordBuilder.build().recordStreamItem().record(), consensusNow);
```
Comment (rbair23) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java:156]: I think we need to add regardless of whether it is a duplicate or not. So you probably need to move the logic in 156 down into the implementation for checkDuplicates, such that we add before we throw in the case of a duplicate, or just add in the case it is not a duplicate.

**Review by rbair23 — CHANGES_REQUESTED** [review]

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/HederaInjectionComponent.java (hunk 6 above)
Comment (vtronkov) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/HederaInjectionComponent.java]: I added them for consistency but you are right - they not used

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/recordcache/RecordCacheService.java [hunk 8]
```diff
@@ -34,7 +34,7 @@ public class RecordCacheService implements Service {
     /** The record cache service name */
     public static final String NAME = "RecordCache";
     /** The name of the queue that stores the transaction records */
-    static final String TXN_RECORD_QUEUE = "TransactionRecordQueue";
+    public static final String TXN_RECORD_QUEUE = "TransactionRecordQueue";
```
Comment (rbair23) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/recordcache/RecordCacheService.java:37]: I'm not sure if this change was needed, did it need to be made public?

**Review by rbair23 — COMMENTED** [review]
Other than the open issue on improving tests, looks good!

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/recordcache/RecordCacheService.java (hunk 8 above)
Comment (vtronkov) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/recordcache/RecordCacheService.java:37]: I probably pushed that with some other changes. Currently, I'm using it on the new test I start writing but it's not pushed yet. I will wait for the story to be finalized so I can finish and push that test.

**Comment by vtronkov** [comment]
> Other than the open issue on improving tests, looks good!

Thank you. Per our Slack group chat, I will wait for Michael's input on how to get the TransactionRecord(since `recordBuilder.build().recordStreamItem().record()` is not the correct way) and then I will finalize and push the tests

**Comment by vtronkov** [comment]
I forgot to turn on the "Sing-off" check in IntelliJ and DCO is failing. I will wait for Michael's PR(https://github.com/hashgraph/hedera-services/pull/7556) to be merged first and then I will follow the instructions under the DCO details to fix it.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/recordcache/RecordCacheService.java (hunk 8 above)
Comment (rbair23) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/recordcache/RecordCacheService.java:37]: Is it necessary to make it public? I fight to keep things as restricted in scope as possible. It helps with readability and safety. A public field might be used by anywhere, so I have to "find usages" to figure out who depends on it and why. If it is package private, I know at least it is only accessed within the package, so I can immediately mentally eliminate most of the code. I stress out changing something from private to package private for the sake of testing (which I hardly ever will do), I would really try very hard to avoid making something public for the sake of testing.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java [hunk 9]
```diff
@@ -201,15 +208,51 @@ private void handlePlatformTransaction(
...
+        if (preHandleResult != null) { // this should always be true, checking just in case
+            try {
+                checkDuplicatesAndIncludeInCache(
+                        preHandleResult,
+                        recordListResult.mainRecord().recordStreamItem().record(),
+                        consensusNow);
+            } catch (final PreCheckException e) {
+                recordFailedTransaction(e.responseCode(), recordBuilder, recordListBuilder);
+            }
+        }
```
Comment (rbair23) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java:225]: The duplicate check is in the wrong place. It **must** happen before we execute the transaction, and before we create the record stream. The purpose of the duplicate check is so that if the same transaction is sent twice, we only execute it once. Otherwise, we will have a double spend problem, which would be very bad. Also, it is a waste of resources to execute a transaction twice. Also, we must know the status of the transaction before we send it to the record manager.

This all needs to move up around line 155 or so. I think what we want to do is:
0. Check whether this transaction is a node due diligence failure. In that case, charge the node and skip to finalization.
1. Check whether this transaction is a duplicate, and add it to the cache.
2. Whether or not it is a duplicate, verify the payer signature.
3. If the payer signature is invalid or the payer has insufficient funds, charge the node.
4. If the payer signature is valid and the payer has sufficient funds, update the record build to indicate this is a duplicate transaction, and charge the user accordingly (finalize transaction, submit record file, etc)
5. If it was not a duplicate, then after verifying the payer signature, execute the transaction, finalize it, etc.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java [hunk 10]
```diff
@@ -201,15 +208,51 @@ private void handlePlatformTransaction(
...
+    }
+
+    private void addTransactionToCache(
+            @NonNull final PreHandleResult preHandleResult,
+            @NonNull final TransactionRecord transactionRecord,
+            @NonNull final Instant consensusNow) {
+        final var txBody = preHandleResult.txInfo().txBody();
+        final var nodeId = txBody.nodeAccountID().accountNum();
+        final var payerAccountId = preHandleResult.payer();
+        hederaRecordCache.add(nodeId, payerAccountId, transactionRecord, consensusNow);
```
Comment (rbair23) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java:252]: My preference is that for a method like this, that is private, and used in only one place, and is only a few lines of code, is to fold the code back into the calling method and include a comment. I know this opinion differs from engineer to engineer, but for me, personally, this kind of method doesn't have enough importance to be broken out. But that's subjective.

**Review by rbair23 — CHANGES_REQUESTED** [review]
The deduplication check must be moved to the right place in the code. Please also enhance the tests so they enforce the order of operations is correct for duplication checks vs. node-due-diligence tests vs. normal transaction execution. Thank you!

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java (hunk 9 above)
Comment (rbair23) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java:225]: Oh! I made a mistake above. Actually, we need to break the duplication check into TWO independent steps. Oof.

It is true that we need to check for duplicates BEFORE we do "handle". The list of steps above are right about that. But, we SHOULD NOT add the transaction to the cache until AFTER we know the final status (about where you have it on line 216).

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java [hunk 11]
```diff
@@ -201,15 +208,51 @@ private void handlePlatformTransaction(
             recordBuilder.status(ResponseCodeEnum.UNKNOWN);
         }
 
-        // TODO: handle long scheduled transactions
+        // store all records at once
+        final var recordListResult = recordListBuilder.build();
+        recordManager.endUserTransaction(recordListResult.recordStream());
+
+        // Verify if the transaction is a duplicate and add it to the cache
+        if (preHandleResult != null) { // this should always be true, checking just in case
```
Comment (rbair23) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java:216]: This is good to check, just in case! However, it raises the question: if we're worried this might happen and we are defending against it, then what do we do if it **IS** null? Do we log it? The transaction would be missing from the record cache, which would mean a replay of the transaction could happen, which would be VERY bad. Maybe it is better to die hard? Maybe we should add to the record cache BEFORE we commit things and write to the record stream so that if this happens, we do not end up committing any changes to state and we do not emit a record.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java [hunk 12]
```diff
@@ -201,15 +208,51 @@ private void handlePlatformTransaction(
...
+            throw new PreCheckException(DUPLICATE_TRANSACTION);
+        }
+    }
+
+    private void addTransactionToCache(
+            @NonNull final PreHandleResult preHandleResult,
+            @NonNull final TransactionRecord transactionRecord,
+            @NonNull final Instant consensusNow) {
+        final var txBody = preHandleResult.txInfo().txBody();
+        final var nodeId = txBody.nodeAccountID().accountNum();
```
Comment (rbair23) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java:250]: The nodeId and the node's account ID are two different things. The node ID uniquely identifies the node over the lifetime of the node, whereas (theoretically) the operator account tied to the node can change over time.

Comment (rbair23) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java:250]: I think we need to add the nodeId to the PreHandleResult. The PreHandleWorkflow knows what the node ID is, so it just needs to stash it away so you can pick it up in the handle workflow.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java (hunk 10 above)
Comment (vtronkov) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java:252]: I really like these kinds of methods because they are easy to understand and are simplifying the main methods. In this case, if I try to find a bug or something in handlePlatformTransaction I might not be interested in the cache at all and I will just skip addTransactionToCache method.

That being said if this is not how it's done in Hedenra I'm okay with refactoring it. It's better to be consistent. Just let me know and I will delete this method.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java (hunk 12 above)
Comment (vtronkov) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java:250]: Sounds good, but I'm not sure how to get the nodeID from the PreHandleWorkflow.
We were looking it with Miro and can we use the networkInfo class to get the nodeID? Something like:
`networkInfo.selfNodeInfo().accountId().accountNum()`

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java (hunk 11 above)
Comment (vtronkov) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java:216]: Maybe intead of depending on the preHandleResult I can get the txBody and payer like we do in PreHandleWorkflowImpl? Something like:
```
final var txBytes = Bytes.wrap(platformTx.getContents());
final TransactionInfo txInfo;
try {
    txInfo = transactionChecker.parseAndCheck(txBytes);
} catch (PreCheckException preCheck) {
    // TODO: call recordFailedTransaction(e.responseCode(), recordBuilder, recordListBuilder); ?
}


final var txBody = txInfo.txBody();
final var payer = txId.accountID();
```
And I can remove the check altogether?

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java (hunk 9 above)
Comment (vtronkov) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java]: Okay, I pushed some changes. I split the cache addition from the duplication check. Now if the transaction is duplicated and signature verification is okay I'm setting the status of the recordBuilder to DUPLICATE_TRANSACTION and I'm calling the `createPenaltyPayment` method and leaving the flow to continue to process the penalty.

When we implement the createPenaltyPayment method(https://github.com/hashgraph/hedera-services/issues/6811) maybe we should have an argument to decide whether to charge the node or the user(or maybe we can have a second method for that).

About point **0** - this is already implemented in getCurrentPreHandleResult method
About point **3** - I'm not sure if this is something I should implement here or if this is for a different story?

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java (hunk 12 above)
Comment (vtronkov) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java:257]: @rbair23 As we discussed in the meeting yesterday, I've tried passing the nodeId to the preHandle but the underlying problem is `HederaWorkflow.getCurrentPreHandleResult()`. Since we don't know if a pre-handle is called before the handle there is no way for me to access the nodeId passed on pre-handle in the case where the pre-handle was not run yet. This is the case after this if statement(`if (preHandleStillValid(configuration, metadata))`) in the getCurrentPreHandleResult method.

I've noticed that we are building the **creator** in `getCurrentPreHandleResult` the same way we build it in `Hedera` when we pass it - `final var creator = networkInfo.nodeInfo(platformEvent.getCreatorId().id());`
If that's the case the `platformEvent.getCreatorId().id()` should be the **nodeId** and I can use that directly in `HandleWorkflow` without the need to pass it from `Hedera`. What do you think?

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java (hunk 10 above)
Comment (vtronkov) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java:259]: The method got reduced so much that it really doesn't make much sense to have it as a separate method. I removed the method 👍

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/ingest/SubmissionManager.java [hunk 13]
```diff
@@ -97,17 +97,16 @@ public SubmissionManager(
             @NonNull final ConfigProvider configProvider,
             @NonNull final Metrics metrics) {
         this.platform = requireNonNull(platform);
-        this.submittedTxns = requireNonNull(deduplicationCache);
```
Comment (netopyr) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/ingest/SubmissionManager.java:100]: I think it is good practice to keep `this.` when a field is initialized in the constructor.

File: hedera-node/hedera-app/src/test/java/com/hedera/node/app/workflows/TransactionScenarioBuilder.java [hunk 14]
```diff
@@ -138,6 +139,8 @@ public static TransactionBody goodDefaultBody() {
                         .build())
                 .transactionFee(1L)
                 .transactionValidDuration(Duration.newBuilder().seconds(60).build())
+                .nodeAccountID(ALICE.accountID())
+                .utilPrng(UtilPrngTransactionBody.newBuilder().range(10).build())
```
Comment (netopyr) [inline hedera-node/hedera-app/src/test/java/com/hedera/node/app/workflows/TransactionScenarioBuilder.java:143]: Why did you add this line? The `TransactionBody` is invalid if both `cryptoTransfer` and `utilPrng` are set.

File: hedera-node/hedera-app/src/test/java/com/hedera/node/app/workflows/TransactionScenarioBuilder.java [hunk 15]
```diff
@@ -138,6 +139,8 @@ public static TransactionBody goodDefaultBody() {
                         .build())
                 .transactionFee(1L)
                 .transactionValidDuration(Duration.newBuilder().seconds(60).build())
+                .nodeAccountID(ALICE.accountID())
```
Comment (netopyr) [inline hedera-node/hedera-app/src/test/java/com/hedera/node/app/workflows/TransactionScenarioBuilder.java:142]: We should use a different `AccountID`, IMO. Otherwise, the payer and the node account are the same, which is not typical.

File: hedera-node/hedera-app/src/test/java/com/hedera/node/app/workflows/handle/HandleWorkflowTest.java [hunk 16]
```diff
@@ -89,14 +104,21 @@ class HandleWorkflowTest extends AppTestBase {
...
+
         final var key = ALICE.account().keyOrThrow();
         return new PreHandleResult(
                 ALICE.accountID(),
                 key,
                 status,
                 code,
-                new TransactionScenarioBuilder().txInfo(),
+                new TransactionScenarioBuilder()
+                        .withTransactionID(transactionId)
```
Comment (netopyr) [inline hedera-node/hedera-app/src/test/java/com/hedera/node/app/workflows/handle/HandleWorkflowTest.java:120]: This should not be necessary. There was a bug in `SingleTransactionRecordBuilder`, which is fixed now.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java [hunk 17]
```diff
@@ -139,19 +146,20 @@ private void handlePlatformTransaction(
...
         // Get the consensus timestamp
-        final Instant consensusNow = platformTxn.getConsensusTimestamp();
+        final var consensusNow = platformTxn.getConsensusTimestamp();
 
         // Setup record builder list
         blockRecordManager.startUserTransaction(consensusNow, state);
         final var recordBuilder = new SingleTransactionRecordBuilder(consensusNow);
         final var recordListBuilder = new RecordListBuilder(recordBuilder);
 
+        PreHandleResult preHandleResult = null;
```
Comment (netopyr) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java:156]: I think the code block below has to be structured differently. It is hard to follow which block is executed under which circumstances. I am actually pretty sure that the current implementation is wrong. (For instance, if `nodeCharge` is true, we want to skip most of the rest.) But it is hard to tell.

File: hedera-node/hedera-app/src/test/java/com/hedera/node/app/workflows/handle/HandleWorkflowTest.java [hunk 18]
```diff
@@ -170,6 +203,12 @@ void setup(@Mock InstantSource instantSource) {
                 .when(dispatcher)
                 .dispatchHandle(any());
 
+        final var fakeHederaState = new FakeHederaState();
+        fakeHederaState.addService(RecordCacheService.NAME, Map.of("TransactionRecordQueue", new LinkedList<>()));
+        when(workingStateAccessor.getHederaState()).thenReturn(fakeHederaState);
+
+        hederaRecordCache = spy(new RecordCacheImpl(deduplicationCache, workingStateAccessor, configProvider));
```
Comment (netopyr) [inline hedera-node/hedera-app/src/test/java/com/hedera/node/app/workflows/handle/HandleWorkflowTest.java:210]: Not part of this PR, but it makes me nervous that we pass `workingStateAccessor` into this constructor. I wouldn't be surprised if this introduces very subtle bugs.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/ingest/SubmissionManager.java (hunk 13 above)
Comment (vtronkov) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/ingest/SubmissionManager.java:100]: Yes...my IntelliJ keeps removing those. I think I fix it and it shouldn't happen now.

File: hedera-node/hedera-app/src/test/java/com/hedera/node/app/workflows/TransactionScenarioBuilder.java (hunk 14 above)
Comment (vtronkov) [reply hedera-node/hedera-app/src/test/java/com/hedera/node/app/workflows/TransactionScenarioBuilder.java:143]: It was related to SingleTransactionRecordBuilder somehow but now it's not needed since there were some changes there and I refactored the tests. Removed 👍

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java (hunk 17 above)
Comment (netopyr) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java:156]: I think before we do a significant refactoring of the workflow, we should update the diagram and review that.

**Comment by vtronkov** [comment]
Closing due to git problems. Here is the new PR: https://github.com/hashgraph/hedera-services/pull/7697
